src/txmempool.cpp

```diff
@@ -854,7 +854,6 @@ void CTxMemPool::PrioritiseTransaction(const uint256& hash, const CAmount& nFeeD
 
 void CTxMemPool::ApplyDelta(const uint256 hash, CAmount &nFeeDelta) const
 {
-    LOCK(cs);
```

Comment:
Here and in `CTxMemPool::ClearPrioritisation()`, `CTxMemPool::GetTotalTxSize()` and `CTxMemPool::IsUnbroadcastTx()` I would strongly advise to add `AssertLockHeld(cs)` where `LOCK(cs)` is removed. And in general in every function that is annotated `EXCLUSIVE_LOCKS_REQUIRED()`. It would serve as documentation to the reader and also ensure that the code does not get executed without holding the mutex.

The clang compile time checks are good, but I would say not to be the sole protection - they get silenced in an unexpected way if `EXCLUSIVE_LOCKS_REQUIRED()` and `ASSERT_EXCLUSIVE_LOCK()` are both used, do not work with gcc and produce only compile warnings with clang (could be missed if -Werror is not used). I would say they are an addition, but not replacement of run-time asserts.

Reply:
6dbd57fca9d324dd468d0ee627392bd788606bd8

> they get silenced in an unexpected way

Do you have a code sample?

> I would say they are an addition

Agree.

Reply:
> > they get silenced in an unexpected way
> 
> @vasild do you have a code sample?

Yes, see https://github.com/bitcoin/bitcoin/pull/19668#discussion_r467045878. There are two unexpected things:

1. The warnings differ depending on the order of the attributes (if more than one is specified):
  ```cpp
  void f() EXCLUSIVE_LOCKS_REQUIRED(cs) ASSERT_EXCLUSIVE_LOCK(cs);
  // vs
  void f() ASSERT_EXCLUSIVE_LOCK(cs) EXCLUSIVE_LOCKS_REQUIRED(cs);
  ```

2. The presence of `ASSERT_EXCLUSIVE_LOCK()` alters the behavior of `EXCLUSIVE_LOCKS_REQUIRED()`.

Reply:
Annotating both `ASSERT_EXCLUSIVE_LOCK` and `EXCLUSIVE_LOCKS_REQUIRED` on the same function makes no sense. The only reason to do the `ASSERT_EXCLUSIVE_LOCK` is if you're relying on the test having succeeded for execution to continue as far as later compile-time checks are concerned, but if you've satisfied `EXCLUSIVE_LOCKS_REQUIRED` already, then there's no need to hint the later compile-time checks, they're already satisfied.

Reply:
> It was tedious to check the callers of e.g. `ApplyDelta()` and their callers and their callers etc until finding either `LOCK(cs)` or `AssertLockHeld(cs)` to ensure that the mutex is indeed held in all places where `ApplyDelta()` is called.

The whole point of the compile time annotations is to have the compiler do the work of checking that the callers have a `LOCK(cs)`. Note that an `AssertLockHeld(cs)` isn't sufficient -- any time that's called without having the lock already held would also be a bug; and until the `ASSERT_EXCLUSIVE_LOCK` annotation is removed from `AssertLockHeld`, `AssertLockHeld` may in fact prevent the compiler from doing adequate checking...